test/extensions/transport_sockets/tls/handshaker_test.cc

```diff
@@ -163,6 +165,83 @@ TEST_F(HandshakerTest, ErrorCbOnAbnormalOperation) {
   EXPECT_EQ(post_io_action, Network::PostIoAction::Close);
 }
 
+// Example HandshakerImpl demonstrating special-case behavior which necessitates
+// extra SSL_ERROR case handling. Here, we induce an SSL_ERROR_WANT_X509_LOOKUP,
+// check for it in the handshaker, faux-trigger the lookup, and then proceed as
+// normal.
+class HandshakerImplForTest : public Ssl::Handshaker {
+public:
+  HandshakerImplForTest(bssl::UniquePtr<SSL> ssl, std::function<void()> requested_cert_cb)
+      : ssl_(std::move(ssl)), requested_cert_cb_(requested_cert_cb) {}
+
+  Network::PostIoAction doHandshake(Ssl::SocketState& state,
+                                    Ssl::HandshakerCallbacks& callbacks) override {
+    ASSERT(state != Ssl::SocketState::HandshakeComplete && state != Ssl::SocketState::ShutdownSent);
+
+    int rc = SSL_do_handshake(ssl());
+    if (rc == 1) {
+      state = Ssl::SocketState::HandshakeComplete;
+      callbacks.OnSuccessCb(ssl());
+      return Network::PostIoAction::Close;
+    } else {
+      switch (SSL_get_error(ssl(), rc)) {
+      case SSL_ERROR_WANT_READ:
+      case SSL_ERROR_WANT_WRITE:
+        return Network::PostIoAction::KeepOpen;
+      case SSL_ERROR_WANT_X509_LOOKUP:
+        // Special case. Once this lookup is requested, we flip the bit and allow
+        // the handshake to proceed.
+        requested_cert_cb_();
+        return Network::PostIoAction::KeepOpen;
+      default:
+        callbacks.OnFailureCb();
+        return Network::PostIoAction::Close;
+      }
+    }
+  }
+
+  void setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) override {
+    transport_socket_callbacks_ = &callbacks;
+  }
+
+  SSL* ssl() override { return ssl_.get(); }
+
+private:
+  bssl::UniquePtr<SSL> ssl_;
+  std::function<void()> requested_cert_cb_;
+  Network::TransportSocketCallbacks* transport_socket_callbacks_{};
+};
+
+TEST_F(HandshakerTest, NormalOperationWithHandshakerImplForTest) {
+  bool cert_cb_ok = false;
+
+  ::testing::MockFunction<void()> requested_cert_cb;
+  EXPECT_CALL(requested_cert_cb, Call).WillOnce([&]() { cert_cb_ok = true; });
+
+  SSL_set_cert_cb(
+      server_ssl_.get(), [](SSL*, void* arg) -> int { return *static_cast<bool*>(arg) ? 1 : -1; },
+      &cert_cb_ok);
+
+  HandshakerImplForTest handshaker(std::move(server_ssl_), requested_cert_cb.AsStdFunction());
+
+  Network::MockTransportSocketCallbacks transport_socket_callbacks;
+  handshaker.setTransportSocketCallbacks(transport_socket_callbacks);
+
+  StrictMock<MockHandshakerCallbacks> callbacks;
+  EXPECT_CALL(callbacks, OnSuccessCb).Times(1);
+
+  auto socket_state = Ssl::SocketState::PreHandshake;
+  auto post_io_action = Network::PostIoAction::KeepOpen; // default enum
+
+  while (post_io_action != Network::PostIoAction::Close) {
```

Comment:
If this were an integration test with a real event loop, and getting the cert was async, what would `HandshakerImplForTest` do from whatever callback it got on completion of async operation (such as completion of an http/grpc request)? It seems like it would want to call `doHandshake()`, but what would it pass for `callbacks`?

Reply:
I think that the async/await operation would be totally contained in the custom HandshakerImpl. Handshaker::doHandshake() has the same semantics as SSL_do_handshake() itself, i.e. the callsite is responsible for calling it over and over until it calls one of the success/failure callbacks. Did I understand this question correctly?

Reply:
My question is regarding the `HandshakerCallbacks&` param. When the operation completes, and the Handshaker presumably calls `SSL_do_handshake()` and gets the return code for it being completed successfully, what now? Did the handshaker cache the `HandshakerCallbacks` (and is that allowed)? If not, how does it tell envoy that the handshake is complete? This may just need some documentation, or it may be that the `HandshakerCallbacks` should be passed to an `initialize()` method or something and documented as having a lifetime >= lifetime of the handshaker (or until successcb or errorcb is called).

Reply:
I'll add documentation for this. I don't think any HandshakerImpl should cache callbacks; Envoy knows that a handshake is complete because onSuccessCb() is called exactly once during the lifetime of Handshaker::doHandshake(). If the Handshaker has to do something async(), it shouldn't store the Callbacks struct so that it can call onSuccessCb() later, it should wait until Handshaker::doHandshake() is next called and then invoke onSuccessCb(). I'll try and make this clearer in the documentation.